File: ccx/clipboard.py
"""Cross-platform clipboard support using only stdlib."""

from __future__ import annotations

import shutil
import subprocess
import sys


def copy_to_clipboard(text: str) -> bool:
    """Copy `text` to the system clipboard. Returns True on success."""
    cmd: list[str] | None = None
    if sys.platform == "darwin":
        cmd = ["pbcopy"]
    elif sys.platform == "win32":
        cmd = ["clip"]
    elif sys.platform.startswith("linux") or sys.platform.startswith("freebsd"):
        if shutil.which("wl-copy"):
            cmd = ["wl-copy"]
        elif shutil.which("xclip"):
            cmd = ["xclip", "-selection", "clipboard"]
        elif shutil.which("xsel"):
            cmd = ["xsel", "--clipboard", "--input"]

    if not cmd:
        return False

    try:
        subprocess.run(cmd, input=text.encode("utf-8"), check=True)
        return True
    except (subprocess.CalledProcessError, FileNotFoundError, OSError):
        return False

File: internal/clipboard/clipboard.go
// Package clipboard provides cross-platform copy-to-clipboard by shelling out
// to the platform's standard clipboard CLI (pbcopy / clip / wl-copy / xclip /
// xsel).
package clipboard

import (
	"errors"
	"os/exec"
	"runtime"
	"strings"
)

// ErrNoClipboard is returned when no supported clipboard tool is found on PATH.
var ErrNoClipboard = errors.New("no clipboard tool found (looked for pbcopy/clip/wl-copy/xclip/xsel)")

// Copy writes text to the system clipboard. Returns ErrNoClipboard if no
// supported tool is available, or the underlying error if the tool fails.
func Copy(text string) error {
	cmd := pickCommand()
	if cmd == nil {
		return ErrNoClipboard
	}
	c := exec.Command(cmd[0], cmd[1:]...)
	c.Stdin = strings.NewReader(text)
	return c.Run()
}

func pickCommand() []string {
	switch runtime.GOOS {
	case "darwin":
		if which("pbcopy") {
			return []string{"pbcopy"}
		}
	case "windows":
		if which("clip") {
			return []string{"clip"}
		}
	case "linux", "freebsd", "openbsd", "netbsd":
		if which("wl-copy") {
			return []string{"wl-copy"}
		}
		if which("xclip") {
			return []string{"xclip", "-selection", "clipboard"}
		}
		if which("xsel") {
			return []string{"xsel", "--clipboard", "--input"}
		}
	}
	return nil
}

func which(name string) bool {
	_, err := exec.LookPath(name)
	return err == nil
}
